internals/overlord/logstate/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,381 @@
+		// Zero the removed element to allow garbage collection.
+		c.entries[0] = entryWithService{}
+		c.entries = c.entries[1:]
+	}
+
+	if len(c.entries) >= cap(c.entries) {
+		// Copy all the elements to the start of the buffer.
+		copy(c.buffer, c.entries)
+
+		// Reset the view into the buffer.
+		c.entries = c.buffer[:len(c.entries):len(c.buffer)]
+
+		// Zero removed elements to allow garbage collection.
+		for i := len(c.entries); i < len(c.buffer); i++ {
+			c.buffer[i] = entryWithService{}
+		}
+	}
+
+	c.entries = append(c.entries, entryWithService{
+		entry:   encodeEntry(entry),
+		service: entry.Service,
+	})
+	return nil
+}
+
+func encodeEntry(entry servicelog.Entry) logRecord {
+	message := strings.TrimSuffix(entry.Message, "\n")
+
+	return logRecord{
+		TimeUnixNano: fmt.Sprintf("%d", entry.Time.UnixNano()),
+		Body:         anyValue{StringValue: &message},
+	}
+}
+
+type payload struct {
+	ResourceLogs []resourceLogs `json:"resourceLogs"`
+}
+
+// Flush sends the buffered logs to the OpenTelemetry collector.
+func (c *Client) Flush(ctx context.Context) error {
+	if len(c.entries) == 0 {
+		return nil // no-op
+	}
+
+	// Group entries by service.
+	serviceBatches := make(map[string][]logRecord)
+	for _, otelEntryWithService := range c.entries {
+		serviceName := otelEntryWithService.service
+		logRecord := otelEntryWithService.entry
+		serviceBatches[serviceName] = append(serviceBatches[serviceName], logRecord)
+	}
+
+	serviceNames := make([]string, 0, len(serviceBatches))
+	for serviceName := range serviceBatches {
+		serviceNames = append(serviceNames, serviceName)
+	}
+	// Sort service names to ensure deterministic order.
+	sort.Strings(serviceNames)
+
+	logs := make([]resourceLogs, 0, len(serviceNames))
+	for _, serviceName := range serviceNames {
+		batch := serviceBatches[serviceName]
+		if len(batch) > 0 {
+			resourceAttributes := c.resourceAttributes[serviceName]
+			resource := resource{
+				Attributes: resourceAttributes,
+			}
+			scope := instrumentationScope{Name: "pebble"}
+			scopeLogs := []scopeLogs{
+				{
+					Scope:      scope,
+					LogRecords: batch,
+				},
+			}
+			logs = append(logs, resourceLogs{
+				Resource:  resource,
+				ScopeLogs: scopeLogs,
+			})
+		}
+	}
+
+	if len(logs) == 0 {
+		return nil
+	}
+
+	payload := payload{
+		ResourceLogs: logs,
+	}
+
+	resp, err := c.sendBatch(ctx, payload)
+	if err != nil {
+		return err
+	}
+
+	return c.handleServerResponse(resp)
+}
+
+func (c *Client) sendBatch(ctx context.Context, payload payload) (*http.Response, error) {
+	jsonData, err := json.Marshal(payload)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling log batch: %v", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", c.options.Location+"/v1/logs", bytes.NewBuffer(jsonData))
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("User-Agent", fmt.Sprintf("pebble/%s", cmd.Version))
```

Comment:
Per earlier discussion, let's avoid the weird dependency from internals to cmd (and avoid hardcoding "pebble" here). Instead, let's add a `UserAgent string` field to the `ClientOptions` struct and have the caller set that to `cmd.ProgramName + "/" + cmd.Version`. An empty string could either mean "don't set User-Agent" or we could default to "pebble".

Reply:
Got it, thanks, I did not fully get the previous comment on the user agent...